### ISSUE TITLE
Basic auth on oauth2 token endpoint sometime does not propagate basic auth header properly

### ISSUE THREAD

**Issue by tomaszbrunarski** [Jeffail/benthos#787]
Hi Jeff, 
I have a situation that my token endpoint is behind basic auth - standard Spring authentication I guess.

Normally I query the endpoint as input and then process it - this works.
What does not work is when I try to push the processed data out or consume data from an endpoint requiering the same oauth2. 
https://login:pass@fdsfsfdsf.requestcatcher.com/ works the first time as the token endpoint but then if I try to do requests in parallel is works half the time.

Below is my processor which tries to get images from the source_url endpint and I can see in the request catcher that every second request for the token has endoced ":" as the basic auth header.
The below code is in a branch so no data get pushed along with the token request. 

Is it possible that there is an issue with the oauth2 provider ? I actualy was able to reproduce this issue with insomnia oauth2 provider. 
`processors:
          - http: # souce url https://someserver/4172f8bd-ef33-5703-917f-e13e36cf28fa
              url: '${! meta("source_url")}'
              parallel: true
              verb: GET
              copy_response_headers: true
              #retries: 0
              timeout: 60s
              oauth2:
                enabled: true
                token_url: https://login:password@fdsfsfdsf.requestcatcher.com/
          - catch:
              - log:
                  message: "${!error()} and meta ${!meta()}"`

**Comment by Jeffail** [comment]
Hey @tomaszbrunarski, are you saying it doesn't happen when `parallel` is set to `false`? Or does this always happen on the second token request? I've flagged this as needs more info as I suspect this will require digging into the oauth2 library being used in order to work out what's happening. It might be that basic auth credentials in the URL isn't actually a supported feature and we'll need to solve this another way.

**Comment by tomaszbrunarski** [comment]
Hi @Jeffail , thanks for getting back to me - this happens every  second (not the first) request regardless of the ```parallel``` flag. I think you are right and that might be skipped for some reason every second request it makes. Also ```batch_as_multipart``` is also set to ```false```
I found it very easy to test by setting the oauth token endpoint to the requestcatcher.com  and then triggering 10 messages to go through the processor. With that its preatty consistent that one request goes through it nicely and the second one comes in with only ":" base64 encoded

**Comment by wirtsi** [comment]
Hey Jeff ... do you suggest we should also raise this with the guys that do the oauth2 library? There are some reported issue in the library that describe issues with basic auth (eg https://github.com/golang/oauth2/pull/476) ... the thing with every second request losing the header, not sure if that would be something that happens in benthos or in their lib.

Do you have any ideas what I could report them?

Bests

**Comment by Jeffail** [comment]
Hey @wirtsi, that doesn't look related. I think the problem is that including basic auth credentials in the oauth2 URL is probably not an explicitly supported feature since the lib has it's own way of specifying basic auth. I think we'll probably need to add new config fields to the oauth2 config schema that allow you to specify them outside the URL. Also it's Ash btw, my username is pretty confusing :stuck_out_tongue:

**Comment by tomaszbrunarski** [comment]
There is a fair bit of APIs that work this way so I assume I will not be the only one having this issue. 
Do you need more info? I would help to solve this but know go lang only to the degree that I can run a ready piece of code or tweak it to my needs.

Side-note - this all makes sens now - your name is Ash - Ash vs evil , evil being the chaos of distributed computing which benthos aims to fix. Makes total sense.

**Comment by Jeffail** [comment]
Actually this looks like it might just be a naming/config confusion, reading through the oauth2 code it appears that basic auth in the request is conflated with the client id and client secret, so your config:

```yaml
oauth2:
  enabled: true
  token_url: https://login:password@fdsfsfdsf.requestcatcher.com/
```

Should actually be:

```yaml
oauth2:
  enabled: true
  token_url: https://fdsfsfdsf.requestcatcher.com/
  client_key: login
  client_secret: password
```

Can you try that and report back? The mechanism for populating basic auth fields in the request is based on the target endpoint as it appears they all implement things in a different way: https://github.com/golang/oauth2/blob/master/internal/token.go#L188, so if the auto detection doesn't work we'll need a config field to manually specify how it should work.

**Comment by tomaszbrunarski** [comment]
@Jeffail yup this works, would be great to see this feature in the docs :) im sure more people will want to use this as java spring provides this authentication method in most cases

**Comment by twmb** [comment]
Closing — the issue was resolved in the comments. Use the client_key and client_secret fields instead of embedding credentials in the token URL.
